library/ziti_ctrl.c

```diff
@@ -0,0 +1,293 @@
+/*
+Copyright 2019 Netfoundry, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+https://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#include <stdlib.h>
+#include <http_parser.h>
+#include "utils.h"
+#include "zt_internal.h"
+#include "strutils.h"
+#include "model.h"
+#include <ziti_ctrl.h>
+#include <uv_mbed/um_http.h>
+
+#define MJSON_API_ONLY
+#include <mjson.h>
+
+
+int code_to_error(const char *code) {
+
+#define CODE_MAP(XX) \
+XX(NO_ROUTABLE_INGRESS_NODES, ZITI_GATEWAY_UNAVAILABLE) \
+XX(NO_EDGE_ROUTERS_AVAILABLE, ZITI_GATEWAY_UNAVAILABLE) \
+XX(INVALID_AUTHENTICATION, ZITI_NOT_AUTHORIZED) \
+XX(REQUIRES_CERT_AUTH, ZITI_NOT_AUTHORIZED)\
+XX(UNAUTHORIZED, ZITI_NOT_AUTHORIZED)\
+XX(INVALID_AUTH, ZITI_NOT_AUTHORIZED)
+
+#define CODE_MATCH(c, err) if (strcmp(code,#c) == 0) return err;
+
+    CODE_MAP(CODE_MATCH)
+
+    ZITI_LOG(WARN, "unmapped error code: %s", code);
+    return ZITI_WTF;
+}
+
+struct ctrl_resp {
+    int status;
+    char *body;
+    size_t received;
+    bool resp_chunked;
+
+    void* (*body_parse_func)(const char*, int);
+    void (*resp_cb)(void*, ziti_error*, void*);
+    void *ctx;
+
+    ziti_controller *ctrl;
+    void (*ctrl_cb)(void*, ziti_error*, struct ctrl_resp*);
+};
+
+static void ctrl_resp_cb(um_http_req_t *req, int code, um_header_list *headers) {
+    struct ctrl_resp *resp = req->data;
+    resp->status = code;
+    um_http_hdr *h;
+    LIST_FOREACH(h, headers, _next) {
+        if (strcasecmp(h->name, "Content-Length") == 0) {
+            resp->body = malloc(atol(h->value));
+            break;
+        }
+        if (strcasecmp(h->name, "Transfer-Encoding") == 0 && strcmp(h->value, "chunked") == 0) {
+            resp->resp_chunked = true;
+            resp->body = malloc(0);
+        }
+    }
+}
+
+static void ctrl_default_cb (void *s, ziti_error *e, struct ctrl_resp *resp) {
+    if (resp->resp_cb) {
+        resp->resp_cb(s, e, resp->ctx);
+    }
+
+    free(resp);
+}
+
+static void ctrl_login_cb(ziti_session *s, ziti_error *e, struct ctrl_resp *resp) {
+    resp->ctrl->session = strdup(s->token);
+    um_http_header(&resp->ctrl->client, "zt-session", s->token);
+    ctrl_default_cb(s, e, resp);
+}
+
+static void ctrl_logout_cb(void *s, ziti_error *e, struct ctrl_resp *resp) {
+    FREE(resp->ctrl->session);
+    um_http_header(&resp->ctrl->client, "zt-session", NULL);
```

Comment:
there's no way to just remove a header? or is setting a header to NULL actually removing it from the client?

Reply:
setting to NULL === remove